composer/trainer/trainer.py

```diff
@@ -1158,6 +1164,36 @@ def __init__(
         # Run Event.INIT
         self.engine.run_event(Event.INIT)
 
+        # If the experiment is being tracked with an `MLFlowLogger`, then MLFlow experiment and run are available
+        # after Event.INIT.
+        if save_folder is not None:
+            mlflow_logger = None
+            for destination in self.logger.destinations:
+                if isinstance(destination, MLFlowLogger):
+                    mlflow_logger = destination
+                    break
+
+            if mlflow_logger is not None:
+                mlflow_experiment_id = mlflow_logger._experiment_id
+                mlflow_run_id = mlflow_logger._run_id
+
+                # Log the experiment and run IDs as hyperparameters.
+                self.logger.log_hyperparameters({
+                    'mlflow_experiment_id': mlflow_experiment_id,
+                    'mlflow_run_id': mlflow_run_id
+                })
+
+                # The save folder and related paths/filenames may contain format placeholders for the MLFlow IDs, so
+                # populate them now.
+                mlflow_format_kwargs = {
+                    MLFLOW_EXPERIMENT_ID_FORMAT_KEY: mlflow_experiment_id,
+                    MLFLOW_RUN_ID_FORMAT_KEY: mlflow_run_id
+                }
+
+                save_folder = partial_format(save_folder, **mlflow_format_kwargs)
+                if latest_remote_file_name is not None:
+                    latest_remote_file_name = partial_format(latest_remote_file_name, **mlflow_format_kwargs)
+
```

Comment:
`save_folder` for MLflow will be something like `dbfs:/databricks/mlflow-tracking/{mlflow_experiment_id}/{mlflow_run_id}/artifacts/`, where `{mlflow_experiment_id}` and `{mlflow_run_id}` will be replaced/formatted at runtime.

When the actual object stores are created through `maybe_create_object_store` and `maybe_create_remote_uploader_downloader`, the object stores will sync to the experiment/run initialized by the `MLFlowLogger`. However, from the perspectives of `Trainer` and `CheckpointSaver`, which is constructed pre INIT, `save_folder` is still the original value with the placeholders. The updates to `save_folder` both here and in `CheckpointSaver` are primarily UX improvements, as logging messages would otherwise continue to use the placeholders rather than the actual paths being used by the underlying object stores.

Trainer autoresume example:
```
2024-01-01 22:39:44,635: rank7[785][MainThread]: DEBUG: composer.trainer.trainer: Read path databricks/mlflow-tracking/{mlflow_experiment_id}/{mlflow_run_id}/artifacts/checkpoints/ift-mpt-7b-zomexk/checkpoints/ep0-ba2 from symlink file
2024-01-01 22:39:44,636: rank7[785][MainThread]: INFO: composer.trainer.trainer: Autoresuming training from checkpoint
```

CheckpointSaver example:
```
024-01-01 22:42:42,820: rank7[785][MainThread]: DEBUG: composer.utils.checkpoint: Saving sharded checkpoints to databricks/mlflow-tracking/{mlflow_experiment_id}/{mlflow_run_id}/artifacts/checkpoints/ift-mpt-7b-zomexk/checkpoints/ep0-ba4/__7_0.distcp...
2024-01-01 22:42:57,619: rank7[785][MainThread]: DEBUG: composer.callbacks.checkpoint_saver: Checkpoint locally saved to databricks/mlflow-tracking/{mlflow_experiment_id}/{mlflow_run_id}/artifacts/checkpoints/ift-mpt-7b-zomexk/checkpoints/ep0-ba4/__7_0.distcp
2024-01-01 22:42:57,619: rank7[785][MainThread]: DEBUG: composer.callbacks.checkpoint_saver: Uploading checkpoint to databricks/mlflow-tracking/{mlflow_experiment_id}/{mlflow_run_id}/artifacts/checkpoints/ift-mpt-7b-zomexk/checkpoints/ep0-ba4/__7_0.distcp
2024-01-01 22:43:05,797: rank7[785][MainThread]: DEBUG: composer.callbacks.checkpoint_saver: Creating symlink file /tmp/tmpyblyc0lg/latest.symlink -> databricks/mlflow-tracking/{mlflow_experiment_id}/{mlflow_run_id}/artifacts/checkpoints/ift-mpt-7b-zomexk/checkpoints/ep0-ba4
```

Reply:
Can we use `self._checkpoint_saver.save_folder`

Reply:
Unfortunately cannot - `CheckpointSaver` is initialized with a different value (`parsed_save_folder`) that only contains the save folder path but not the rest of the remote URI information.